enhancements/sig-architecture/30-clusterset-override/README.md

```diff
@@ -0,0 +1,279 @@
+# ManagedClusterSet Override
+
+## Release Signoff Checklist
+ 
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [website](https://github.com/open-cluster-management/website/)
+ 
+## Summary
+The proposed work enhances the managedClusterSet API to support managedClusterSet Override.
+
+## Motivation
+Today, managedClusterSet is designed to group the managedClusters, a managedCluster can only be included in only one managedClusterSet, by setting a label `cluster.open-cluster-management.io/clusterset` on the managedCluster.
+A user must have explicit RBAC rules, `managedclustersets/join`, to add such labels.
+There are also raising requirements on using managedClusterSet for the following purposes. Within a purpose, the managedClusterSet must be exclusive, but for different purposes it can overlap.
+
+- As cluster admin, I want to create a managedClusterSet which includes all clusters in the apac region. Then I can apply certain configurations to all clusters in apac.
+- As a SRE team member, I want to use a managedClusterSet for Disaster Recovery purposes. The managedClusterSet should select two or more clusters in different regions. 
+- As a DEV team member, I want to use a managedClusterSet to select clusters based on the middleware enabled on these clusters, so I could run special applications in these clusters.
+
+## Proposal
+
+### Changes on managedClusterSet APIs.
+
+Currently, managedClusterSets use the label `cluster.open-cluster-management.io/clusterset:<ManagedClusterSet Name>` to select managedClusters. 
+But for some use cases, I want to use other labels to select managedClusters in a managedClusterSets. Meanwhile, managedClusterSets should be exclusive within a purpose, but for different purposes it can overlap.
+So, In this proposal, managedClusterSets could use any labels to select managedClusters. 
+
+So, we add a field `ClusterLabel` into managedClusterSet spec
+
+```go
+ type ManagedClusterSetSpec struct {
+ // ClusterLabel defines the label which the managedClusterSet will use to select managedClusters.
+ // If the field is not set, we will use "cluster.open-cluster-management.io/clusterset:<ManagedClusterSet Name>" as default value.
+ // +optional
+ ClusterLabel ClusterLabel  `json:"clusterLabel"`
```

Comment:
part of me predicts this will not be the last grouping mechanism we choose.  How about


```go
type ManagedClusterSetSpec struct {
    Selector ManagedClusterSetSelector
}

type ManagedClusterSetSelector struct{
    // "" means to use the current mechanism of matching by name
    // "LabelSelector" means to use the label 
    // A theoretical "Named" means to use "Named []string" where the managedclustername is set
    SelectorType string 

    LabelSelector *metav1.LabelSelector

}

Reply:
@deads2k Yes, you are right. I'd like to add another selector type `ClusterLabel` to
- Guarantee clustersets with same cluster label key are exclusive;
- Enable additional permission check when cluster joining/leaving a clusterset (the label key should start with the reserved prefix);

```go
type ManagedClusterSetSelector struct{
    // "" means to use the current mechanism of matching by name
    // "ClusterLabel" means to use a particular cluster label. It is guaranteed that clustersets with same label key 
    // are exclusive with each others;  
    // "LabelSelector" means to use the label 
    // A theoretical "Named" means to use "Named []string" where the managedclustername is set
    SelectorType string 

    LabelSelector *metav1.LabelSelector

    ClusterLabel *ClusterLabel
}

//ClusterLabel defines one cluster label
type ClusterLabel struct {
	Key string `json:"key"`
	Value string `json:"value"`
}
``` 

What do you think?

Reply:
> What do you think?

I think that looks good.

Reply:
add the ClusterLabel selector.